### PR TITLE
Remove need for wrapper node with ReactTransitionGroup

### PR DESCRIPTION
**Pull request by jdlehman**
- Allow `ReactCSSTransitionGroup` to take an optional `component` prop
  (defaults to `span`)
- Remove need for wrapper node with `ReactTransitionGroup`.
  - If no children are specified return `null`.
  - If there is one child, return that child without a wrapper node (unless the
    `component` prop was explicitly provided).
  - If there are multiple children, use the provided `component` prop as a
    wrapper node, defaulting to `span` if one was not provided.

### CODE REVIEW

**Comment by rayshan** [comment]
+1

**Comment by gaearon** [comment]
Can you provide some more context on why you see this beneficial in apps? What problems does this solve? Would that be a breaking change, and if so, why is it justified?

**Comment by jdlehman** [comment]
Thanks for taking a look at this @gaearon!

## Context

The primary benefit of this change is getting rid of unnecessary elements from the DOM tree. Other than reducing clutter and the depth of the DOM tree, this aids in the styling components, especially with regards to layout. Take the following example:

```js
<ParentComponent>
  <ReactTransitionGroup {...someProps}>
    <MyComponent />
  </ReactTransitionGroup>
</ParentComponent>
```

The above will produce the following:

```html
<ParentComponent>
  <span>
    <MyComponent />
  </span>
</ParentComponent>
```

If `MyComponent`'s styling/layout is dependent on its context (like if `ParentComponent` uses Flexbox via `display: flex;`), the layout will now be broken implicitly because React always adds a span around children of `ReactTransitionGroup` if the `component` prop is not specified. We would now have to style the span such that the context is passed down and requires extra work and complexity to get our styles to display as intended. Because we essentially want the wrapper to act like Firefox's experimental [`display: contents`](https://samrueby.com/2015/02/09/firefox-is-releasing-support-for-css-display-contents/).

With this PR, the implicit span is omitted if there is only a single child in the `ReactTransitionGroup`, which would help prevent any additional work in the case that we have contextual styling (typically this would be layout related like Flexbox styles etc).

This is an intermediate step in getting rid of the wrapper requirement with `ReactTransitionGroup`, as it makes the wrapper optional for zero or one children. Getting rid of the requirement for a wrapper altogether is probably related to a much larger stream of work like https://github.com/facebook/react/issues/2127.

In the large React application I am building at work, we have a lot of unneeded wrapper components being generated by `ReactTransitionGroup`, and since we are using Flexbox to create our layouts, we have to add classes that do nothing more than pass the Flexbox context through to its children. This change would help reduce complexity in how we style our app, and make our layouts less fragile.

## Is this a breaking change?

This would be a breaking change as it could potentially affect styling/layout for people who did not explicitly define the `component` prop for `ReactTransitionGroup` AND only have a single child being rendered. That said, this change would not require much work to migrate (for the cases that it does affect). Users could just add the explicit `component="span"` prop for places they need it. Though it is simple fix, it definitely needs to be noted in the change log and release notes.

## Related work

We did something similar in [`React responsive`](https://github.com/contra/react-responsive/pull/30). This change would also be helpful for libraries like [velocity-react](https://github.com/twitter-fabric/velocity-react).

## tldr

This change removes the need for a wrapper in `ReactTransitionGroup` in cases where React does not require it in order to render. By doing so we can reduce depth and unnecessary nesting in the DOM tree, which is also helpful when styling, creating layouts, or simply grokking the DOM.

**Comment by rayshan** [comment]
@jdlehman beat me to it. I have the same need - remove unnecessary wrapper elements as they take up resources and may create additional CSS styling work. I made a video:
https://www.youtube.com/watch?v=S4dd4SiKzkE

If you take a look at the DOM tree on the right, you'll see the `div` highlighted, which doesn't have any attributes, holding the `ReactTransitionGroup`. Its attribute is never changed. Only its children's attributes are. There's no need for the wrapper `div` to be there. Now imagine this is Instagram and we're animating every image in a grid, we'll end up with a lot of unnecessary wrapper elements.

**Comment by gaearon** [comment]
What if there is one current item, and then we add a second one. Wouldn’t switching to `span`s while the first one is displayed cause the first item to lose its state?

**Comment by rayshan** [comment]
That's a good point... If we have a single child, currently the outgoing and incoming child's elements already exist in the DOM simultaneously, we should be ok. If we're going from a single child to multiple children, we'll have the problem. Would resolution to #2127 help?

**Comment by gaearon** [comment]
>If we're going from a single child to multiple children, we'll have the problem.

This is the main reason I’m inclined not to take this. Losing the state and changing DOM structure during a very common transition (most lists may sometimes contain one element) is problematic. 

I think that if you want to have full control over the rendered children you might be better off with something like [react-motion](https://github.com/chenglou/react-motion/) which provides you that by letting you “specify” the rendering function as a `children` prop so you have full control. I wonder if we could do something similar?

```js
return (
  <ReactTransitionGroup render={children => children}>
    <MyComponent />
  </ReactTransitionGroup>
)
```

By default `render` prop could be `children => React.createElement(this.props.component, { children })`.

Would this satisfy your use case?

**Comment by gaearon** [comment]
>Would resolution to #2127 help?

Yes but my gut feeling is that by the time it is implemented we’ll have something better than `ReactTransitionGroup` for animations. I wouldn’t count on fragments coming any time soon.

**Comment by rayshan** [comment]
> `<ReactTransitionGroup render={children => children}>`

That's interesting. I think that would work but we should definitely get more folks to chime in. I need to take a closer look at `react-motion`, not familiar with it.

> I wouldn’t count on fragments coming any time soon.

I hear you, and that's ok. Maybe migrating to `react-motion` is the long-term solution, but it'll take me and perhaps other developers a while to switch from orchestrating animation using timing to using physics.

**Comment by gaearon** [comment]
Do I understand correctly that you always use TransitionGroup for a single element only, and that your use case is not so much a “group” but as a way to animate mounting and unmounting of a single component?

**Comment by jdlehman** [comment]
Yes, the issue is for a single element only since there is nothing we can currently do to get rid of the wrapper when there are multiple elements. We make use of `velocity-react`, which uses `ReactTransitionGroup`, to animate both single elements, and multiple elements throughout our app. We did try migrating to `react-motion` a little while back, but I remember that the performance was worse than velocity to the point that it was not usable on lower end Android devices, which we target.

Switching to `react-motion` might be a valid solution for some and I can definitely try it again, but I think performance might be an issue still.

With regards to the concern of adding components on future renders, I think adding a `render` prop that takes a function should be able to satisfy my use case. If we like that approach I can definitely make changes to this PR.

**Comment by rayshan** [comment]
> always use TransitionGroup for a single element only

Sometimes but not always. In the video I posted above it is indeed always a single element upon mount / unmount.

**Comment by jdlehman** [comment]
I opened #6361 based on the previously mentioned idea of allowing in a custom `render` prop, instead of case specific logic.

**Comment by gaearon** [comment]
I’m closing because, per the reasons above, this particular API is inherently problematic.
Let’s continue the discussion about this use case in #6361. If it isn’t fruitful we’ll need to create an issue to track this.

**Comment by gaearon** [comment]
For posterity, it seems that this was achievable with the current API as well: https://github.com/facebook/react/pull/6361#issuecomment-202587067
